lfshook.go

```diff
@@ -16,42 +18,107 @@ var txtFormatter = &logrus.TextFormatter{DisableColors: true}
 // Multiple levels may share a file, but multiple files may not be used for one level
 type PathMap map[logrus.Level]string
 
+// Alternatively map a log level to an io.Writer
+type WriterMap map[logrus.Level]io.Writer
+
 // Hook to handle writing to local log files.
 type lfsHook struct {
-	paths  PathMap
-	levels []logrus.Level
-	lock   *sync.Mutex
+	paths     PathMap
+	writer    WriterMap
+	levels    []logrus.Level
+	lock      *sync.Mutex
 	formatter logrus.Formatter
 }
 
 // Given a map with keys equal to log levels.
 // We can generate our levels handled on the fly, and write to a specific file for each level.
 // We can also write to the same file for all levels. They just need to be specified.
-func NewHook(levelMap PathMap) *lfsHook {
+func NewHook(levelMap interface{}) *lfsHook {
 	hook := &lfsHook{
-		paths: levelMap,
-		lock:  new(sync.Mutex),
+		lock:      new(sync.Mutex),
 		formatter: txtFormatter,
 	}
-	for level, _ := range levelMap {
-		hook.levels = append(hook.levels, level)
+
+	switch levelMap.(type) {
+	case PathMap:
+		hook.paths = levelMap.(PathMap)
+		for level := range levelMap.(PathMap) {
+			hook.levels = append(hook.levels, level)
+		}
+		break
+	case WriterMap:
+		hook.writer = levelMap.(WriterMap)
+		for level := range levelMap.(WriterMap) {
+			hook.levels = append(hook.levels, level)
+		}
+		break
```

Comment:
I guess that just leaves logging the error, but since we log it later it would be redundant...

My thought on panic here is that if they pass an interface that is neither a WriterMap nor a PathMap then they have a code problem that needs to be corrected before releasing. If it were a matter of passing something in that MIGHT be the wrong type that is being programmatically defined elsewhere then I'd agree that `panic` is too severe, but since this is part of the logging infrastructure then I wouldn't think that the type of the map would be getting programmatically generated but would be defined in script.

Hard to say, though. I'm only thinking from the handful of usecases I've had for this hook. InvalidType error seems panic-worthy to me.